### PR TITLE
Enable more browser tests under chrome on circleci

### PR DESCRIPTION
**Pull request by sbc100**

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -208,11 +208,11 @@ jobs:
           command: |
             # --no-sandbox becasue we are running as root and chrome requires
             # this flag for now: https://crbug.com/638180
-            CHROME_FLAGS_BASE="--no-first-run -start-maximized --no-sandbox"
+            CHROME_FLAGS_BASE="--no-first-run -start-maximized --no-sandbox --use-gl=swiftshader"
```
Comment (kripken) [inline .circleci/config.yml]: With firefox we use gallium softpipe to render, and it seems to do all the GL tests properly. Could the same work here, and allow fewer tests to be disabled? Or is swiftshader necessary for some reason?

File: tests/test_browser.py [hunk 2]
```diff
@@ -73,6 +73,25 @@ def shell_with_script(shell_file, output_file, replacement):
       output.write(input.read().replace('{{{ SCRIPT }}}', replacement))
 
 
+def is_chrome():
+  return EMTEST_BROWSER and 'chrom' in EMTEST_BROWSER.lower()
```
Comment (kripken) [inline tests/test_browser.py]: why without the `e`?

File: tests/test_browser.py [hunk 3]
```diff
@@ -1850,15 +1869,23 @@ def test_matrix_identity(self):
     self.btest('gl_matrix_identity.c', expected=['-1882984448', '460451840', '1588195328'], args=['-s', 'LEGACY_GL_EMULATION=1', '-lGL', '-lSDL'])
 
   @requires_graphics_hardware
+  @no_swiftshader
   def test_cubegeom_pre(self):
     self.btest('cubegeom_pre.c', reference='cubegeom_pre.png', args=['-s', 'LEGACY_GL_EMULATION=1', '-lGL', '-lSDL'])
+
+  @requires_graphics_hardware
+  @no_chrome()
```
Comment (kripken) [inline tests/test_browser.py]: no need for two `no_chrome` lines

File: tests/test_browser.py [hunk 4]
```diff
@@ -1850,15 +1869,23 @@ def test_matrix_identity(self):
...
+  @no_chrome("RELOCATABLE=1 forces synchronous compilation which chromedoesn't support")
```
Comment (kripken) [inline tests/test_browser.py]: space after `chrome`.

File: tests/test_browser.py [hunk 5]
```diff
@@ -3046,7 +3076,11 @@ def test_emterpreter_async(self):
       self.btest('emterpreter_async.cpp', '1', args=['-s', 'EMTERPRETIFY=1', '-s', 'EMTERPRETIFY_ASYNC=1', '-O' + str(opts), '-g2'])
 
   def test_emterpreter_async_2(self):
-    self.btest('emterpreter_async_2.cpp', '40', args=['-s', 'EMTERPRETIFY=1', '-s', 'EMTERPRETIFY_ASYNC=1', '-O3'])
+    # Error.stackTraceLimit default to 10-in chrome by this test relies on more
+    # than 40 stack frames being reported.
+    with open('pre.js', 'w') as f:
+      f.write('Error.stackTraceLimit = 80;\n')
+    self.btest('emterpreter_async_2.cpp', '40', args=['-s', 'EMTERPRETIFY=1', '-s', 'EMTERPRETIFY_ASYNC=1', '--pre-js', 'pre.js'])
```
Comment (kripken) [inline tests/test_browser.py]: why was the `-O3` removed?

File: .circleci/config.yml (hunk 1 above)
Comment (sbc100) [reply .circleci/config.yml]: Yes, we should probably switch using the gallium softpipe.   The doesn't work with xvfb though sadly, as I discovered yesterday.   So we can switch away from xvfb as a seperate change lager perhaps?   I the mean time there is no harm in tagging the tests and don't work with switftshader.  Its probably worth opening bug reports for them too.

File: tests/test_browser.py (hunk 2 above)
Comment (sbc100) [reply tests/test_browser.py]: I think its to support "chromium" , this line of code is not new, I just moved it.

File: tests/test_browser.py (hunk 3 above)
Comment (sbc100) [reply tests/test_browser.py]: Done

File: tests/test_browser.py (hunk 5 above)
Comment (sbc100) [reply tests/test_browser.py]: oops, reverted that part.

File: tests/test_browser.py [hunk 6]
```diff
@@ -73,6 +73,25 @@ def shell_with_script(shell_file, output_file, replacement):
...
+def no_chrome(note='chome is not supported'):
+  if is_chrome():
+    return unittest.skip(note)
+  return lambda f: f
+
+
+def no_swiftshader(f):
+  def decorated(self):
+    if is_chrome() and '--use-gl=swiftshader' in EMTEST_BROWSER:
+      return self.skipTest('not compatible with swiftshader')
```
Comment (jgravelle-google) [inline tests/test_browser.py]: Do we need to return skipTest? I thought it threw an exception

File: tests/test_browser.py [hunk 7]
```diff
@@ -73,6 +73,25 @@ def shell_with_script(shell_file, output_file, replacement):
...
+    return f(self)
```
Comment (jgravelle-google) [inline tests/test_browser.py]: Ditto do we need to return f(self), or just run it?

File: tests/test_browser.py [hunk 8]
```diff
@@ -3046,7 +3075,12 @@ def test_emterpreter_async(self):
       self.btest('emterpreter_async.cpp', '1', args=['-s', 'EMTERPRETIFY=1', '-s', 'EMTERPRETIFY_ASYNC=1', '-O' + str(opts), '-g2'])
 
   def test_emterpreter_async_2(self):
-    self.btest('emterpreter_async_2.cpp', '40', args=['-s', 'EMTERPRETIFY=1', '-s', 'EMTERPRETIFY_ASYNC=1', '-O3'])
+    # Error.stackTraceLimit default to 10 in chrome by this test relies on more
```
Comment (jgravelle-google) [inline tests/test_browser.py]: "default to 10 in chrome by this test" is confusing to me.

Oh, typo: s/by/but/

File: tests/test_browser.py [hunk 9]
```diff
@@ -3372,10 +3408,9 @@ def test_pthread_cancel(self):
     self.btest(path_from_root('tests', 'pthread', 'test_pthread_cancel.cpp'), expected='1', args=['-O3', '-s', 'USE_PTHREADS=1', '-s', 'PTHREAD_POOL_SIZE=8'])
 
   # Test pthread_kill() operation
+  # This test hangs the chrome render process, and keep subsequent tests from passing too
```
Comment (jgravelle-google) [inline tests/test_browser.py]: can/should this comment be folded in to the no_chrome note? Seems almost redundant to have both

**Pull request merged**